### PR TITLE
allow for serialization of the histogram into raw and base64 formats

### PR DESCRIPTION
**Pull request by husobee**
This will convert the histogram into a a representation consumable by the write histogram api.  New public methods added are SerializeRaw and SerializeB64.

### CODE REVIEW

**Comment by HeinrichHartmann** [comment]
We have implementations of libcircllhist in C and JS as well. Public methods should be consistent between the implementations.

File: circonusllhist.go [hunk 1]
```diff
@@ -238,6 +241,149 @@ type Histogram struct {
...
+		count |= buf[int(offset)+int(i)+3] << (uint8(i) * 8)
+	}
+	bucket.count = uint64(count)
+
+	h.insertBin(&bucket, int64(count))
+
+	return uint8(offset) + 3 + uint8(tgtType) + 1, nil
+}
+
+func (h *Histogram) bvWrite(idx int) []uint8 {
```
Comment (sanjayar) [inline circonusllhist.go]: could you add a comment that describes this encoding.

File: circonusllhist.go [hunk 2]
```diff
@@ -238,6 +241,149 @@ type Histogram struct {
...
+	result[2] = uint8(tgtType)
+	for i := tgtType; i >= 0; i-- {
+		var count = uint8(h.bvs[idx].count)
+		result[i+3] = ((count >> (uint8(i) * 8)) & 0xff)
+	}
+
+	return result
+}
+
+func htons(b *[]byte, i, v uint8) {
```
Comment (sanjayar) [inline circonusllhist.go]: these operations are available from encoding.binary.BigEndian

**Comment by husobee** [comment]
@HeinrichHartmann: OK, So instead of `h.SerializeRaw` and `h.SerializeB64` they should be module functions `HistSerialize(histogram, buffer, length)` and `HistDeserializeB64(histogram, buffer, length)`?  Same with `HistDeserialize(histogram, buffer, length)` and `HistDeserializeB64`?  Thanks for the suggestion!

**Comment by HeinrichHartmann** [comment]
Well, I am not 100% sure if there is need for change. Personally I am fine with methods like `h.SerializeB64` instead of `HistSerializeB64(h)`.

I was more concerned about the functions `SerializeRaw`/`DeserializeRaw`,`bvWrite`/`bvRead` that don't have a parallel C version. Are those publicly exposed?

File: internal_test.go [hunk 3]
```diff
@@ -1,10 +1,35 @@
 package circonusllhist
```
Comment (sanjayar) [inline internal_test.go]: why isnt this file circonusllhist_test.go?

File: internal_test.go [hunk 4]
```diff
@@ -1,10 +1,35 @@
...
 
 import (
+	"bytes"
 	"math"
 	"testing"
 )
```
Comment (sanjayar) [inline internal_test.go]: there really ought to be a test that decodes and encodes a binary buffer right?

**Comment by husobee** [comment]
bvRead and bvWrite are not publically exposed.  I will update SerializeRaw to just Serialize to match the naming conventions.  Thanks!  Pushing very soon.

File: circonusllhist.go [hunk 5]
```diff
@@ -238,6 +241,149 @@ type Histogram struct {
...
+	BVL2 = 1
+	BVL3 = 2
+	BVL4 = 3
+	BVL5 = 4
+	BVL6 = 5
+	BVL7 = 6
+	BVL8 = 7
+)
+
+func (h *Histogram) bvRead(buf []byte, offset int) (uint8, error) {
```
Comment (sanjayar) [inline circonusllhist.go]: if you named argument 1 it would be clearer that its a size

File: circonusllhist.go [hunk 6]
```diff
@@ -238,6 +241,149 @@ type Histogram struct {
...
+	if h.bvs == nil {
+		h.bvs = make([]bin, defaultHistSize)
+	}
+
+	if len(buf) < 3 {
+		return 0, errors.New("buf length must be greater than 3")
+	}
+	var tgtType = buf[offset+2]
+
+	if tgtType > BVL8 {
```
Comment (sanjayar) [inline circonusllhist.go]: this message could be more direct and include the actual offending value.

File: circonusllhist.go [hunk 7]
```diff
@@ -238,6 +241,149 @@ type Histogram struct {
...
+		return 0, errors.New("value in buf is greater than BVL8")
+	}
+
+	if len(buf) < int(3+tgtType+1) {
```
Comment (sanjayar) [inline circonusllhist.go]: might as well include the expected vs got sizes here too

File: internal_test.go (hunk 3 above)
Comment (husobee) [reply internal_test.go]: renamed in latest commit to circonusllhist_test.go

**Pull request merged**